cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java

```diff
@@ -448,14 +436,33 @@ public void apply() throws Exception {
   }
 
   // Returns true if the upgrade flag is set. Upgrade could have completed earlier than this since this flag is
-  // updated asynchronously.
+  // updated using a cache loader with a fixed timeout.
   public boolean isUpgradeComplete() {
     if (upgradeCacheLoader == null) {
       return false;
     }
     return upgradeCacheLoader.getUnchecked(APP_VERSION_UPGRADE_KEY);
   }
 
+  private synchronized boolean executeUpgradeInTransaction(final Table table, final AtomicInteger maxNumberUpgradeRows)
```

Comment:
Is the TransactionExecutorFactory thread safe? The other methods that were accessing it were synchronized.